### PR TITLE
Added support for Accept-Encoding header

### PR DESCRIPTION
**Pull request by Pask423**

### CODE REVIEW

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package sttp.model.headers
+
+import sttp.model.internal.Validate.RichEither
+
+case class AcceptEncoding(compressionAlgorithm: String, weight: Option[BigDecimal]) {
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:5]: According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding I think the right name for `compressionAlgorithm` would be `contentEncoding`

Also, there might be multiple content encodings specified, each with an optional weight (and I think that's often the case). So maybe sth like `case class AcceptEncoding(contentEncodings: List[Weighted[String]])`?

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+  def unsafeParse(s: String): List[AcceptEncoding] = parse(s).map(_.getOrThrow)
+
+  def unsafeApply(compressionAlgorithm: String, weight: Option[BigDecimal]): AcceptEncoding =
+    safeApply(compressionAlgorithm, weight).getOrThrow
+
+  def safeApply(compressionAlgorithm: String, weight: Option[BigDecimal]): Either[String, AcceptEncoding] = {
+    val encoding = AcceptEncoding(compressionAlgorithm, weight)
+    if (isValid(encoding)) Right(encoding)
+    else Left("Invalid Content Range")
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:51]: invalid encoding? or do we in fact validate the weight?

File: README.md [hunk 3]
```diff
@@ -36,6 +36,7 @@ Available model classes include:
    * `ServerSentEvent`  
 * header values:
    * `Accepts`
+   * `Accept-Encoding`
```
Comment (adamw) [inline README.md:39]: We should also add a factory method to `Header` which accepts the `AcceptEncoding` value

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
+package sttp.model.headers
+
+import sttp.model.internal.Validate.RichEither
+
+import scala.annotation.tailrec
+
+case class Encoding(encoding: String, weight: Option[BigDecimal]) {
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:7]: I think I'd move it inside the `AcceptEncoding` object - or will it be used by also header values as well? If not, then the weighted-header seems like a detail of the accept encoding header, not a "global" concept

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  override def toString: String = encodings.map(_.toString).mkString(",")
+}
+
+object AcceptEncoding {
+
+  def parse(str: String): Either[String, AcceptEncoding] = {
+    val encodings = processString(str, List.empty)
+    if (encodings.forall(isValid) && encodings.nonEmpty) Right(AcceptEncoding(encodings.reverse))
+    else if (encodings.isEmpty)
+      Left("Expected Accept-Encoding in the format: \"deflate or gzip;q=1.0\", but got: %s".format(str))
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:22]: you can't have an `or` in the header value, right? ;)

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        weight.split("=") match {
+          case Array(_, value) => Encoding(algorithm, Some(BigDecimal(value)))
+          case _               => Encoding("", None)
+        }
+      case Array(algorithm) => Encoding(algorithm, None)
+      case _                => Encoding("", None)
+    }
+  }
+
+  private def isValid(acceptEncoding: Encoding): Boolean = {
```
Comment (adamw) [inline core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:49]: maybe we can return a more descriptive error message saying what's wrong?

File: core/src/main/scala/sttp/model/headers/AcceptEncoding.scala (hunk 5 above)
Comment (Pask423) [reply core/src/main/scala/sttp/model/headers/AcceptEncoding.scala:22]: yeah should fix this

**Pull request merged**